src/main/java/org/apache/spark/shuffle/ucx/reducer/UcxShuffleClient.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (C) Mellanox Technologies Ltd. 2019. ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+package org.apache.spark.shuffle.ucx.reducer;
+
+import org.apache.spark.SparkEnv;
+import org.apache.spark.executor.TempShuffleReadMetrics;
+import org.apache.spark.network.shuffle.BlockFetchingListener;
+import org.apache.spark.network.shuffle.DownloadFileManager;
+import org.apache.spark.network.shuffle.ShuffleClient;
+import org.apache.spark.shuffle.ShuffleHandle;
+import org.apache.spark.shuffle.UcxShuffleManager;
+import org.apache.spark.shuffle.UcxWorkerWrapper;
+import org.apache.spark.shuffle.ucx.memory.MemoryPool;
+import org.apache.spark.shuffle.ucx.memory.RegisteredMemory;
+import org.apache.spark.storage.BlockId$;
+import org.apache.spark.storage.BlockManagerId;
+import org.apache.spark.storage.ShuffleBlockId;
+import org.openucx.jucx.UcxUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.openucx.jucx.UcxException;
+import org.openucx.jucx.ucp.UcpEndpoint;
+import org.openucx.jucx.ucp.UcpRemoteKey;
+import scala.Option;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.HashMap;
+
+public class UcxShuffleClient extends ShuffleClient {
+  private final MemoryPool mempool;
+  private static final Logger logger = LoggerFactory.getLogger(UcxShuffleClient.class);
+  private final UcxShuffleManager ucxShuffleManager;
+  private final ShuffleHandle handle;
+  private final TempShuffleReadMetrics shuffleReadMetrics;
+  private final UcxWorkerWrapper workerWrapper;
+  final HashMap<Integer, UcpRemoteKey> offsetRkeysCache = new HashMap<>();
+  final HashMap<Integer, UcpRemoteKey> dataRkeysCache = new HashMap<>();
+
+  public UcxShuffleClient(ShuffleHandle handle,
+                          TempShuffleReadMetrics shuffleReadMetrics,
+                          UcxWorkerWrapper workerWrapper) {
+    this.ucxShuffleManager = (UcxShuffleManager) SparkEnv.get().shuffleManager();
+    this.mempool = ucxShuffleManager.ucxNode().getMemoryPool();
+    this.handle = handle;
+    this.shuffleReadMetrics = shuffleReadMetrics;
+    this.workerWrapper = workerWrapper;
+  }
+
+  /**
+   * Submits n non blocking fetch offsets to get needed offsets for n blocks.
+   */
+  private void submitFetchOffsets(UcpEndpoint endpoint, ShuffleBlockId[] blockIds,
+                                  long[] dataAddresses, RegisteredMemory offsetMemory) {
+    ByteBuffer driverMetadata =  workerWrapper.fetchDriverMetadataBuffer(handle.shuffleId()).data();
```

Comment:
maybe wrap parsing the driver metadata of a specific mapper id to a separate class to separate parsing from index fetch action